testsuite/oiiotool-demosaic/run.py

```diff
@@ -10,75 +10,109 @@
 width = 160
 height = 120
 
-# oiiotool commands to make each of the Bayer patterns
-make_pattern_RGGB = f"--pattern constant:color=0,1,0 {width}x{height} 3 -for y \"0,{{TOP.height}},2\" -for x \"0,{{TOP.width}},2\" -point:color=1,0,0 \"{{x}},{{y}}\" -point:color=0,0,1 \"{{x+1}},{{y+1}}\" -endfor -endfor"
-make_pattern_GRBG = f"--pattern constant:color=0,1,0 {width}x{height} 3 -for y \"0,{{TOP.height}},2\" -for x \"1,{{TOP.width}},2\" -point:color=1,0,0 \"{{x}},{{y}}\" -point:color=0,0,1 \"{{x-1}},{{y+1}}\" -endfor -endfor"
-make_pattern_GBRG = f"--pattern constant:color=0,1,0 {width}x{height} 3 -for y \"1,{{TOP.height}},2\" -for x \"0,{{TOP.width}},2\" -point:color=1,0,0 \"{{x}},{{y}}\" -point:color=0,0,1 \"{{x+1}},{{y-1}}\" -endfor -endfor"
-make_pattern_BGGR = f"--pattern constant:color=0,1,0 {width}x{height} 3 -for y \"1,{{TOP.height}},2\" -for x \"1,{{TOP.width}},2\" -point:color=1,0,0 \"{{x}},{{y}}\" -point:color=0,0,1 \"{{x-1}},{{y-1}}\" -endfor -endfor"
-
-layouts = {
-    "RGGB": make_pattern_RGGB,
-    "GRBG": make_pattern_GRBG,
-    "GBRG": make_pattern_GBRG,
-    "BGGR": make_pattern_BGGR
-}
-
-types = [
-    {
-        "type" : "float",
-        "ext" : "exr",
-        "threshold": 0.0000003
-    },
-    {
-        "type" : "half",
-        "ext" : "exr",
-        "threshold": 0.0005
-    },
+# Create a test image with color gradient
+make_testimage = f"--pattern fill:topleft=0,0,1:topright=0,1,0:bottomleft=1,0,1:bottomright=1,1,0 {width}x{height} 3 "
```

Comment:
This run.py has changed a lot. I'm not entirely sure I understand how it works. I totally trust you on this, but I urge you to run this test in isolation and make sure it doesn't seem like it takes too long. There was a point there I visited this test not long ago because it was extraordinarily expensive, noticed a ton of redundant work being done, rewrote it a bit, and sped up by 3x (see #4429). I just want to make sure we aren't altering it in a way that goes back to this test being disproportionately expensive versus the narrow feature it's trying to test.

Reply:
It works exactly the same way as before, I just changed the code to generate the 4 layouts (RGGB, GRBG, GBRG, BGGR) programmatically, using a nested 2x2 loop. That reduces the code duplication, mostly in the part where we generate the bayer patterns. Since some tests now require un-whitebalanced images, there would have been even more code duplication without that.

The code still uses your trick with pre-generating images. It works slightly slower, as it writes more files, but it doesn't write more files than necessary. Runs pretty fast for me.